### ISSUE TITLE
instrumentation issue when using private constructors

### ISSUE THREAD

**Issue by rvowles** [ebean-orm/ebean#1751]
GITHUB ISSUES ARE STRICTLY CONTROLLED FOR THIS PROJECT.

Refer to http://ebean-orm.github.io/support for the policies controlling the use of github issues.
Please post issues to the Ebean group https://groups.google.com/forum/#!forum/ebean first.

## Expected behavior

ebean instrumentation doesn't explode. 

## Actual behavior

ebean instrumentation explodes in IDE. It generally seems to figure it out with the in-maven instrumentation, but in IDE with the old (compile time) and new (runtime) instrumenter it always explodes.

The hope is it would just ignore private constructors or those that take parameters. the Builder is a static class in the same class.

### Steps to reproduce
(using the InnerBuilder from IDEA's plugin list)

```java
  public DbGroup() {}

  private DbGroup(Builder builder) {
    name = builder.name;
  }
```

```console
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    io/X/db/model/DbGroup.<init>(Lio/X/db/model/DbGroup$Builder;)V @58: invokevirtual
  Reason:
    Type 'io/X/db/model/DbGroup$Builder' (current frame, stack[1]) is not assignable to 'io/X/db/model/DbGroup'
  Current Frame:
    bci: @58
    flags: { }
    locals: { 'io/X/db/model/DbGroup', 'io/X/db/model/DbGroup$Builder' }
    stack: { 'io/X/db/model/DbGroup', 'io/X/db/model/DbGroup$Builder' }
  Bytecode:
    0000000: 2ab7 0048 2abb 004a 592a b700 4db5 004f
    0000010: 2a2b b400 54b6 0058 2a2b b400 5ab6 005e
    0000020: 2a2b b400 60b6 0064 2a2b b400 66b6 006a
    0000030: 2a2b b400 6cb6 0070 2a2b b600 74b6 0078
    0000040: 2a2b b600 7bb6 007e b1                 

	at java.lang.Class.getDeclaredMethods0(Native Method) ~[?:?]
	at java.lang.Class.privateGetDeclaredMethods(Class.java:3167) ~[?:?]
	at java.lang.Class.privateGetPublicMethods(Class.java:3192) ~[?:?]
	at java.lang.Class.getMethods(Class.java:1905) ~[?:?]
	at io.ebeaninternal.server.deploy.BeanLifecycleAdapterFactory.addLifecycleMethods(BeanLifecycleAdapterFactory.java:46) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.deploy.BeanDescriptorManager.createDeployBeanInfo(BeanDescriptorManager.java:1317) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.deploy.BeanDescriptorManager.readEntityDeploymentInitial(BeanDescriptorManager.java:715) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.deploy.BeanDescriptorManager.deploy(BeanDescriptorManager.java:371) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.core.InternalConfiguration.<init>(InternalConfiguration.java:197) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.core.DefaultContainer.createServer(DefaultContainer.java:124) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.core.DefaultContainer.createServer(DefaultContainer.java:69) ~[ebean-11.39.1.jar:?]
	at io.ebeaninternal.server.core.DefaultContainer.createServer(DefaultContainer.java:35) ~[ebean-11.39.1.jar:?]
	at io.ebean.EbeanServerFactory.create(EbeanServerFactory.java:58) ~[ebean-11.39.1.jar:?]
	at io.ebean.Ebean$ServerManager.getWithCreate(Ebean.java:126) ~[ebean-11.39.1.jar:?]
	at io.ebean.Ebean$ServerManager.<init>(Ebean.java:75) [ebean-11.39.1.jar:?]
	at io.ebean.Ebean$ServerManager.<init>(Ebean.java:48) [ebean-11.39.1.jar:?]
	at io.ebean.Ebean.<clinit>(Ebean.java:43) [ebean-11.39.1.jar:?]
	at io.ebean.DB.getDefault(DB.java:76) [ebean-11.39.1.jar:?]
```

**Comment by rbygrave** [comment]
Can you provide the full class for DbGroup?  I don't see an '@entity' annotation for example so this class would not be enhanced.  I won't be able to reproduce with the above example?

Cheers,Rob.

**Comment by rvowles** [comment]
relevant bits instead of whole thing?

```
@Entity
@Table(name = "fh_group")
public class DbGroup {

  @Id
  private UUID id;

  public DbGroup() {}

  private DbGroup(Builder builder) {
    setWhoCreated(builder.whoCreated);
    setOwningPortfolio(builder.owningPortfolio);
    setAdminGroup(builder.adminGroup);
    setOwningOrganization(builder.owningOrganization);
    setName(builder.name);
    setGroupRolesAcl(builder.groupRolesAcl);
    setPeopleInGroup(builder.peopleInGroup);
  }

...

public static final class Builder {
    private DbPerson whoCreated;
    private DbPortfolio owningPortfolio;
    private boolean adminGroup;
    private DbOrganization owningOrganization;
    private String name;
    private Set<DbAcl> groupRolesAcl;
    private Set<DbPerson> peopleInGroup;

    public Builder() {
    }

    public Builder whoCreated(DbPerson val) {
      whoCreated = val;
      return this;
    }

    public Builder owningPortfolio(DbPortfolio val) {
      owningPortfolio = val;
      return this;
    }

    public Builder adminGroup(boolean val) {
      adminGroup = val;
      return this;
    }

    public Builder owningOrganization(DbOrganization val) {
      owningOrganization = val;
      return this;
    }

    public Builder name(String val) {
      name = val;
      return this;
    }

    public Builder groupRolesAcl(Set<DbAcl> val) {
      groupRolesAcl = val;
      return this;
    }

    public Builder peopleInGroup(Set<DbPerson> val) {
      peopleInGroup = val;
      return this;
    }

    public DbGroup build() {
      return new DbGroup(this);
    }
  }
}
```

**Comment by rbygrave** [comment]
That might be enough.  If I copy the above I presume that will reproduce the issue.

Btw: technically our entity beans don't actually need setters or getters.  Those fluent builder methods could be on the entity bean itself if desired.

**Comment by rbygrave** [comment]
@rvowles  I was unable to reproduce:   Refer to this branch https://github.com/ebean-orm/ebean-agent/compare/feature/ebean_1751?expand=1

**Comment by rbygrave** [comment]
I have reproduced this with Java 11 compilation.

**Comment by rbygrave** [comment]
So the problem here in the constructor that is taking the builder is that ... 
```
 GETFIELD test/model/domain/DbGroup$Builder.peopleInGroup : Ljava/util/Set;
```
The above GETFIELD on the builder was getting intercepted / changed. The reason is because there is a bug in the constructor enhancement visitor (that isn't in the method enhancement visitor) ... that the `constructor enhancement visitor` wasn't checking that the GETFIELD is on the type of the entity being enhanced (it enhanced `DbGroup$Builder.peopleInGroup` because it is the same field name `peopleInGroup` of the entity field but forgot to check that the field is in fact owned by `DbGroup$Builder` so it should not intercept it at all.

**Comment by rbygrave** [comment]
Fixed by ebean-agent version 11.41.2 - https://github.com/ebean-orm/ebean-agent/issues/119
